ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Forge.hs

```diff
@@ -97,17 +97,20 @@ evolveKESKeyIfNecessary updateForgeState (SL.KESPeriod kesPeriod) = do
   where
     -- | Evolve the given key so that its KES period matches @kesPeriod@.
     evolveKey :: HotKey c -> HotKey c
-    evolveKey (HotKey oldPeriod outdatedKey) = go outdatedKey oldPeriod kesPeriod
-      where
-        go !sk c t
-          | t < c
-          = error "Asked to evolve KES key to old period"
-          | c == t
-          = HotKey kesPeriod sk
-          | otherwise
-          = case KES.updateKES () sk c of
-              Nothing  -> error "Could not update KES key"
-              Just sk' -> go sk' (c + 1) t
+    evolveKey !hk@(HotKey kesPeriodOfKey key)
+        | kesPeriod == kesPeriodOfKey
+        = hk
+        | kesPeriod < kesPeriodOfKey
+        = error $
+            "Asked to evolve KES signature key at period " <>
+            show kesPeriodOfKey <> "  to old period " <> show kesPeriod
+        | otherwise
+        = case KES.updateKES () key kesPeriodOfKey of
+            Just key' -> evolveKey (HotKey (kesPeriodOfKey + 1) key')
+            Nothing   -> error $
+              "Could not evolve KES signature key to period " <>
+              show kesPeriod <> ", its current period " <>
+              show kesPeriodOfKey <> " is its last"
```

Comment:
An alternative is to throw a new type of exception for this, but the annoying thing is that we'd need to reintroduce `RunNode.nodeExceptionIsFatal` (removed in #2243) so that we can mark this exception as fatal in the `ErrorPolicy`. The goal is to empty out `RunNode` :slightly_frowning_face:.

Hmm, actually, exceptions thrown by `error` are not marked as fatal in the error policy (not our choice). So how come the current `error` is causing the node to shut down (see https://github.com/input-output-hk/cardano-node/issues/1235) :thinking: